ambuda/views/auth.py

```diff
@@ -105,14 +164,75 @@ def sign_out():
     return redirect(url_for(POST_AUTH_ROUTE))
 
 
-@bp.route("/recover", methods=["GET", "POST"])
-def recover():
-    form = RecoverForm()
+@bp.route("/reset-password", methods=["GET", "POST"])
+def get_reset_password_token():
+    """Email the user a password reset link."""
+    form = ResetPasswordForm()
     if form.validate_on_submit():
-        hashed = generate_password_hash(form.password.data)
-    else:
-        hashed = ""
-    return render_template("auth/recover.html", form=form, hashed=hashed)
+        email = form.email.data
+        session = q.get_session()
+        user = session.query(db.User).filter_by(email=email).first()
+        if user:
+            raw_token = _create_reset_token(user.id)
+            mail.send_reset_password_email(
+                username=user.username, email=user.email, raw_token=raw_token
+            )
+            return render_template("auth/reset-password-post.html", email=user.email)
+        else:
+            flash(
+                "Sorry, the email address you provided is not associated with any of our acounts."
+            )
+
+    # Override the default message ("The response parameter is missing.")
+    # for better UX.
+    if form.recaptcha.errors:
+        form.recaptcha.errors = ["Please click the reCAPTCHA box."]
+
+    return render_template("auth/reset-password.html", form=form)
+
+
+@bp.route("/reset-password/<username>/<raw_token>", methods=["GET", "POST"])
+def reset_password_from_token(username, raw_token):
+    """Reset password after the user clicks a reset link."""
+    msg_invalid = "Sorry, this reset password link isn't valid. Please try again."
+
+    user = q.user(username)
+    if user is None:
+        flash(msg_invalid)
+        return redirect(url_for("auth.get_reset_password_token"))
+
+    token = _get_reset_token_for_user(user.id)
+    if not _is_valid_reset_token(token, raw_token):
+        flash(msg_invalid)
+        return redirect(url_for("auth.get_reset_password_token"))
+
+    form = ResetPasswordFromTokenForm()
+    if form.validate_on_submit():
+        has_password_match = form.password.data == form.confirm_password.data
+        has_token_match = form.raw_token.data == raw_token
+        if has_password_match and has_token_match:
+            user.set_password(form.password.data)
+            login_user(user, remember=True)
```

Comment:
Maybe it's better not to automatically log in the user, but instead redirect to sign-in? This way the user is forced to sign in again and potentially re-save their password with their password manager (if it wasn't already saved in the reset password form step)

Reply:
From my quick spelunks it seems like a gray area. There's one camp that fiercely says that they don't want to enter their password in a third time given that they just provided it. There's another that says it's useful to ensure that you really know the password and it also assists password managers.

The [OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Forgot_Password_Cheat_Sheet.html) recommendation is:

> Don't automatically log the user in, as this introduces additional complexity to the authentication and session handling code, and increases the likelihood of introducing vulnerabilities.

So I'll just follow that.